README

gbrainy README

Please send questions and suggestions to
http://live.gnome.org/gbrainy

gbrainy is licensed under the GPL, see COPYING for more details.

----

gbrainy is a brain teaser game and trainer to have fun and to keep your brain trained.

It provides the following types of games:

* Logic puzzles. Games designed to challenge your reasoning and thinking skills.

* Mental calculation. Games based on arithmetical operations designed to prove your mental calculation skills.

* Memory trainers. Games designed to challenge your short term memory.

----

gbrainy homepage is located at http://live.gnome.org/gbrainy.

gbrainy is stored in Gnome SVN, in the gbriany repository. 
See http://svn.gnome.org/viewcvs/gbrainy/

If you're interested in discussing gbrainy, consider the gbrainy mailing list.
To subscribe, visit
http://groups.google.com/group/gbrainy


NEWS
gbrainy NEWS

Version 2.1.3
* Updated translations & fixes
* 1 bug fix

Version 2.1.2
* Updated translations
* 8 bug fixes

Version 2.1.1
* Updated translations
* 1 bug fix

Version 2.1.0
* Software improvements
** Refactoring to break down responsibilities in classes and make code more modular
** Automated “Assembly / package versioning”
* Existing games improvements
** Mouse support for Memorize figures and text
** Show graphically the answer in square sheets puzzle
** Add rationale to handshake puzzle
** People at the table has a better graphic
** Some controls in games handle better larger texts
* Translations
** New Telugu translation
** Many updated translations

Version 2.06
* 1 bug fix
* Updated Spanish

Version 2.05
* 1 bug fix
* Updated translations

Version 2.04
* 3 bug fixes
* Fixes to translations

Version 2.03
* Several translations updated to fix bugs
* Add Esperanto translation
* Fixes assembly versioning

Version 2.02
* 7 bugs fixed
* Updated translations for many languages

Version 2.01
* Add Mono.Addins.Gui 0.6 strings for L10N
* remove libgnome support (deprecated lib)
* 11 bugs fixed

Version 2.00
* UI
** Integrated question and solution
** Theme support
** New default theme (notebook)
** Vertical / Horizontal toolbar orientation
* Game experience
** Better answer handling
** Accept words, accept figures names, more flexibility handling spaces, etc
** Better rationale explanations (%)
** Fixed some verbal analogies
* Localization
** I18N fixes
*** Proof-reading on English by native speakers. 20% of strings where enhanced for better grammar
**** Ability to run gbrainy in English ignoring the translations 
* 11 new games: 4 logic games, 1 calculation game, 6 new verbal analogies
* Infrastructure Fixes
** Minor memory leaks
** Mono 2.10 support 
* Several bug fixes

Version 1.65
* 2 bug fixes
* Updated translations

Version 1.64
* 2 bug fixes
* Updated French translation

Version 1.63
* 2 bug fixes
* Updated and new translations
* More flexibility accepting answers

Version 1.62
* 4 bug fixes
* Updated and new translations

Version 1.61
* 5 bug fixes
* Updated and new translations

Version 1.60
* Better I18N (string concatenation and plurals)
* 7 new logic games, 1 calculation and 16 verbal analogies
* PDF export support
* 14 bug fixes
* Updated and new translations

Version 1.52
* 3 bug fixes
* Updated and new translations

Version 1.51
* 5 bug fixes
* Updated and new translations

Version 1.50
* 11 new logic games, 4 new calculation trainers, 13 verbal analogies
* Usability enhancements from Alexandre Laplante and Jeff Stewart, students from University of Ottawa
* Unit testing for the gbrainy.Core
* Ability to define external logic puzzles using an xml file (games.xml)
* Command line support
* 7 bug fixes

Version 1.42
* 14 bug fixes
* Updated and new translations
* Updated Mallard documentation

Version 1.41
* 18 bug fixes
* Updated and new translations

Version 1.4
* 2 new logic games
* Deprecated Glade and moved to Builder
* Mouse support
* Better RTL support
* Updated and new translations
* Remove dead and legacy code
* 12 Bug fixes 

Version 1.3
* Re-architectured for better multiplatform support
* Personal records tracking
* Improved player scoring algorithm
* Better accessibility
* 9 new verbal analogies and a new logic puzzle
* Mallard documentation
* More HIG complaint 
* Bug fixes

Version 1.2
* 6 new games
* Verbal analogies
* New translations
* Bug fixes

Version 1.1
* 4 new games
* Use Mono.Addins for extending gbrainy
* New translations
* Bug fixes

Version 1.00
* 1 new logic puzzle, 1 new memory puzzle
* Better look and feel (new background)
* Switch font drawing to Pango
* Better I18N support
* Bug fixes

Version 0.7
* 8 new puzzles
* License included in the about box
* The drawing area is now square and the application resizes better
* Better score feedback
* Player's history
* Preferences persistence
* Better game selection
* Bug fixes

Version 0.6
* Implemented difficulty levels: easy, medium and master
* Complete revision of the texts
* 1 new math game and 3 new logic games
* Bug fixes
 
Version 0.5
* Basic user preferences (memory games)
* 4 new logic games and 2 new memory trainers
* Count down for memory games (to get your memory ready)
* Some bug fixes

Version 0.4
* Many bugs fixes and corrections
* More tip and better explanations
* New Welcome screen that gives a better explanation of what gbrainy is about
* A complete score system and summary of the score
* A toolbar that makes more accessible the main options
* 4 new logic games, 1 new math trainer and 1 memory trainer

Version 0.3
* Many bugs fixes and corrections
* Many questions and answers are more precise
* Custom game support: define which puzzles you want to play
* 3 new logic puzzles: Divide circles, matrix groups, Balance
* 2 new math puzzles: Greater divisor, two numbers
* Fixes in gbrainy.glade to make it work on Win32
* Small UI enhancements

Version 0.2
* Fixed make distcheck 
* 3 new logic Puzzles (Cube, sheets, figures and words)
* Refactor GameManager, GameSession and other classes
* 4 new memory trainers (numbers, colours, figures, words)
* 1 new math trainer (arithmetic)
* New application icon
* Fixed a couple of usability bugs

Version 0.1
* Initial release
* 13 different logic puzzles included
* Possibility to localize it to different languages (Catalan included)



src/Game.cs
/*
 * Copyright (C) 2007 Jordi Mas i Hernàndez <jmas@softcatala.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

using System;
using Mono.Unix;

abstract public class Game
{
	// See: GetGameTypeDescription
	public enum Types
	{	
		None			= 0,
		LogicPuzzle		= 2,
		MemoryTrainer		= 4,
		MathTrainer		= 8,
		VerbalAnalogy		= 16,
	}

	public enum Difficulty
	{
		None			= 0,
		Easy			= 2,
		Medium			= 4,
		Master			= 8,
	}

	private bool draw_answer;
	private gbrainy application;
	private Cairo.Color default_color;
	protected string right_answer;
	protected Random random;
	private TimeSpan game_time;
	private bool won;
	private bool tip_used;
	private Difficulty difficulty;

	protected Game ()
	{
		random = new Random ();
		draw_answer = false;
		default_color = new Cairo.Color (0, 0, 0);
		won = false;
		tip_used = false;
		difficulty = Difficulty.Medium;
	}

	public abstract string Question {
		get;
	}

	public virtual string Answer {
		get {
			return String.Format (Catalog.GetString ("The correct answer is {0}."), right_answer);
		}
	}

	// Stores how difficult the game is
	public virtual Difficulty GameDifficulty {
		get {
			return Difficulty.Master | Difficulty.Medium | Difficulty.Easy;
		}
	}

	// The level of difficulty selected for the current game
	public virtual Difficulty CurrentDifficulty {
		set {
			difficulty = value;
		}
		get {
			return difficulty;
		}
	}

	public abstract string Name {
		get;
	}

	public string TipString {
		get { 
			string tip = Tip;
	
			if (tip != string.Empty)
				tip_used = true;

			return tip;
		}
	}

	public virtual string Tip {
		get { return string.Empty;}
	}
	
	public gbrainy App {
		get {return application; }
		set {application = value; }
	}

	public virtual bool ButtonsActive {
		get { return true;}
	}

	public virtual Types Type {
		get { return Types.LogicPuzzle;}
	}

	public bool DrawAnswer {
		get { return draw_answer; }
		set { draw_answer = value; }
	}

	// An initialized game cannot be playable (for example, missing external files)
	public virtual bool IsPlayable {
		get { return true;}
	}

	public virtual double DrawAreaX {
		get {return 0.1;}
	}

	public virtual double DrawAreaY {
		get {return 0.1;}
	}

	public virtual double DrawAreaWidth {
		get {return 1 - DrawAreaX * 2;}
	}

	public virtual double DrawAreaHeight {
		get {return 1 - DrawAreaY * 2;}
	}

	public virtual double LineWidth {
		get {return 0.005; }
	}

	public virtual Cairo.Color DefaultDrawingColor {
		get {return default_color; }
	}

	public TimeSpan GameTime {
		get {return game_time; }
		set {game_time = value; }
	}

	public bool Won {
		get { return won; }
		set { won = value; }
	}

	// Average time in seconds that a player is expected to complete this game
	public int AverageTime {
		get {
			double factor;

			switch (CurrentDifficulty) {
			case Difficulty.Easy:
				factor = 1.3;
				break;
			case Difficulty.Master:
				factor = 0.7;
				break;		
			case Difficulty.Medium:
			default:
				factor = 1.0;
				break;		
			}
				
			switch (Type) {
			case Types.MemoryTrainer:
				return (int) (30 * factor);
			case Types.MathTrainer:
				return (int) (60 * factor);
			}
			return (int) (120 * factor); // Default for all games (logic)
		}
	}

	//
	// Score algorithm return a value between 0 and 10
	//
	public virtual int Score {
		get {
			double score;
			double seconds = GameTime.TotalSeconds;

			if (won == false) {
				score = 0;
			} else {		
				score = 10;
			
				// Time
				if (seconds > AverageTime * 3) {
					score = score * 0.6;
				}
				else if (seconds > AverageTime * 2) {
					score = score * 0.7;
				} else if (seconds > AverageTime) {
					score = score * 0.8;
				}
		
				if (tip_used) {
					score = score * 0.8;
				}
			}

			return (int) score;
		}
	}
	
	public abstract void Initialize ();
	public virtual void Finish () {}

	static public string GetPossibleAnswer (int answer)
	{
		switch (answer) {
			// Translators Note
			// The following series of answers may need to be adapted
			// in cultures with alphabets different to the Latin one.
			// The idea is to enumerate a sequence of possible answers
			// For languages represented with the Latin alphabet use 
			// the same than English
		case 0: // First possible answer for a series (e.g.: Figure A)
			return Catalog.GetString ("A");
		case 1: // Second possible answer for a series
			return Catalog.GetString ("B");
		case 2: // Third possible answer for a series
			return Catalog.GetString ("C");
		case 3: // Fourth possible answer for a series
			return Catalog.GetString ("D");
		case 4: // Fifth possible answer for a series
			return Catalog.GetString ("E");
		case 5: // Sixth possible answer for a series
			return Catalog.GetString ("F");
		case 6: // Seventh possible answer for a series
			return Catalog.GetString ("G");
		case 7: // Eighth possible answer for a series
			return Catalog.GetString ("H");
		default:
			return string.Empty;
		}
	}

	public string GetPossibleFigureAnswer (int answer)
	{
		return String.Format (Catalog.GetString ("Figure {0}"), GetPossibleAnswer (answer));
	}

	public virtual void Draw (CairoContextEx gr, int width, int height)
	{
		gr.Scale (width, height);
		gr.DrawBackground ();
		gr.Color = new Cairo.Color (0, 0, 0);
		gr.LineWidth = LineWidth;
	}

	public virtual void DrawPreview (CairoContextEx gr, int width, int height)
	{
		Draw (gr, width, height);
	}

	public virtual bool CheckAnswer (string answer)
	{
		return (String.Compare (answer, right_answer, true) == 0);
	}

	// When asking for a list of figures people trends to use spaces or commas
	// to separate the elements
	static public string TrimAnswer (string answer)
	{
		string rslt = string.Empty;

		for (int i = 0; i < answer.Length; i++)
		{
			if (answer[i]==' ' || answer[i] == ',')
				continue;

			rslt += answer[i];
		}
		return rslt;
	}

	// Type enum to string representation
	static public string GetGameTypeDescription (Types type)
	{
		string str;

		switch (type) 
		{
			case Game.Types.LogicPuzzle:
				str = Catalog.GetString ("Logic");
				break;
			case Game.Types.MemoryTrainer:
				str = Catalog.GetString ("Memory");
				break;
			case Game.Types.MathTrainer:
				str = Catalog.GetString ("Mental Calculation");
				break;
			case Game.Types.VerbalAnalogy:
				str = Catalog.GetString ("Verbal");
				break;
			default:
				str = string.Empty;
				break;
		}
		return str;
	}
}

